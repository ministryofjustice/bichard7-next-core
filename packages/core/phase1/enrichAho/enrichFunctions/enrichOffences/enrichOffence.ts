import type { OffenceCode } from "@moj-bichard7-developers/bichard7-next-data/dist/types/types"
import type { Offence } from "../../../../types/AnnotatedHearingOutcome"

const clearData = (offence: Offence): void => {
  offence.OffenceCategory = undefined
  offence.RecordableOnPNCindicator = undefined
  offence.NotifiableToHOindicator = undefined
  offence.HomeOfficeClassification = undefined
  offence.ResultHalfLifeHours = undefined
}

const enrichOffence = (
  offence: Offence,
  offenceIgnored: boolean,
  offenceCodeLookup: OffenceCode | undefined
): Offence => {
  clearData(offence)
  const reason = offence.CriminalProsecutionReference.OffenceReason
  if (!reason) {
    return offence
  }

  if (offenceIgnored) {
    offence.OffenceCategory = "B7"
    offence.RecordableOnPNCindicator = false
  } else if (offenceCodeLookup?.offenceCategory) {
    offence.OffenceCategory = offenceCodeLookup.offenceCategory
    offence.RecordableOnPNCindicator = offenceCodeLookup.recordableOnPnc
  }

  if (offenceCodeLookup) {
    if (!offence.OffenceTitle) {
      offence.OffenceTitle = offenceCodeLookup.offenceTitle
    }

    offence.NotifiableToHOindicator = offenceCodeLookup.notifiableToHo
    offence.HomeOfficeClassification = offenceCodeLookup.homeOfficeClassification
  }

  return offence
}

export default enrichOffence
