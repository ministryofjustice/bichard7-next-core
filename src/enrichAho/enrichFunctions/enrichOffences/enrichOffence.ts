import type { OffenceCode } from "bichard7-next-data-latest/dist/types/types"
import type { Offence } from "src/types/AnnotatedHearingOutcome"

const enrichOffence = (
  offence: Offence,
  offenceIgnored: boolean,
  offenceCodeLookup: OffenceCode | undefined
): Offence => {
  const reason = offence.CriminalProsecutionReference.OffenceReason
  if (!reason) {
    return offence
  }

  if (offenceIgnored) {
    offence.OffenceCategory = "B7"
    offence.RecordableOnPNCindicator = false
  } else if (offenceCodeLookup?.offenceCategory) {
    offence.OffenceCategory = offenceCodeLookup.offenceCategory
    offence.RecordableOnPNCindicator = offenceCodeLookup.recordableOnPnc
  }

  if (offenceCodeLookup) {
    offence.OffenceTitle = offenceCodeLookup.offenceTitle
    offence.NotifiableToHOindicator = offenceCodeLookup.notifiableToHo
    offence.HomeOfficeClassification = offenceCodeLookup.homeOfficeClassification
    if (offenceCodeLookup.resultHalfLifeHours) {
      offence.ResultHalfLifeHours = parseInt(offenceCodeLookup.resultHalfLifeHours, 10)
    }
  }

  return offence
}

export default enrichOffence
