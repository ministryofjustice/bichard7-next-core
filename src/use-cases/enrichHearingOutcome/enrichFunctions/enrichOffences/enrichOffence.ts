import type { Offence } from "src/types/AnnotatedHearingOutcome"
import { lookupOffenceByCjsCode } from "src/use-cases/dataLookup"
import isOffenceIgnored from "src/use-cases/isOffenceIgnored"
import getOffenceCode from "src/utils/offence/getOffenceCode"

const enrichOffence = (offence: Offence): Offence => {
  const reason = offence.CriminalProsecutionReference.OffenceReason
  if (!reason) {
    return offence
  }

  const offenceCode = getOffenceCode(offence)
  const offenceIgnored = isOffenceIgnored(offence)
  const offenceCodeLookup = offenceCode ? lookupOffenceByCjsCode(offenceCode) : undefined

  if (offenceIgnored) {
    offence.OffenceCategory = "B7"
    offence.RecordableOnPNCindicator = false
  } else if (offenceCodeLookup?.offenceCategory) {
    offence.OffenceCategory = offenceCodeLookup.offenceCategory
    offence.RecordableOnPNCindicator = offenceCodeLookup.recordableOnPnc
  }

  if (offenceCodeLookup) {
    offence.OffenceTitle = offenceCodeLookup.offenceTitle
    offence.NotifiableToHOindicator = offenceCodeLookup.notifiableToHo
    offence.HomeOfficeClassification = offenceCodeLookup.homeOfficeClassification
    if (offenceCodeLookup.resultHalfLifeHours) {
      offence.ResultHalfLifeHours = parseInt(offenceCodeLookup.resultHalfLifeHours, 10)
    }
  }

  return offence
}

export default enrichOffence
