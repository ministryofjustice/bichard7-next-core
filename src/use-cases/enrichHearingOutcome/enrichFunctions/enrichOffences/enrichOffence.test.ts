import type { Offence } from "src/types/AnnotatedHearingOutcome"
import enrichOffence from "./enrichOffence"

describe("enrichOffence()", () => {
  it("should add the data from the looked up offence", () => {
    const offence = {
      CriminalProsecutionReference: {
        OffenceReason: {
          __type: "LocalOffenceReason",
          LocalOffenceCode: { AreaCode: "05", OffenceCode: "BG73006" }
        }
      }
    } as Offence
    const result = enrichOffence(offence)
    expect(result).toMatchSnapshot()
  })

  it("should add an offence category of B7 if the offence is ignored", () => {
    const offence = {
      CriminalProsecutionReference: {
        OffenceReason: {
          __type: "LocalOffenceReason",
          LocalOffenceCode: { AreaCode: "05", OffenceCode: "05MC001" }
        }
      }
    } as Offence
    const result = enrichOffence(offence)
    expect(result.OffenceCategory).toBe("B7")
  })

  it("should set recordable to false if the offence is ignored", () => {
    const offence = {
      CriminalProsecutionReference: {
        OffenceReason: {
          __type: "LocalOffenceReason",
          LocalOffenceCode: { AreaCode: "05", OffenceCode: "05MC001" }
        }
      },
      RecordableOnPNCindicator: "Y"
    } as Offence
    const result = enrichOffence(offence)
    expect(result.RecordableOnPNCindicator).toBe("N")
  })
})
