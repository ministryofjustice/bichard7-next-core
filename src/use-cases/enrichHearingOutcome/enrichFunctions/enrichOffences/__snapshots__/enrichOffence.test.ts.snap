// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enrichOffence() should add the data from the looked up offence 1`] = `
Object {
  "CriminalProsecutionReference": Object {
    "OffenceReason": Object {
      "LocalOffenceCode": Object {
        "AreaCode": "05",
        "OffenceCode": "BG73006",
      },
      "__type": "LocalOffenceReason",
    },
  },
  "HomeOfficeClassification": "000/00",
  "NotifiableToHOindicator": "N",
  "OffenceCategory": "CE",
  "OffenceTitle": "POSSESSING THING DERIVED FROM DEAD BADGER",
  "RecordableOnPNCindicator": "N",
}
`;
